[WebGPU] Switching pipelines does not zero out dynamic offsets from prior pipeline

mwyrzykowski opened the pull request:
#### 46c6dd5414b60c3c64c27c688ad9b625321007c4
<pre>
[WebGPU] Switching pipelines does not zero out dynamic offsets from prior pipeline
<a href="https://bugs.webkit.org/show_bug.cgi?id=285828">https://bugs.webkit.org/show_bug.cgi?id=285828</a>
<a href="https://rdar.apple.com/142761614">rdar://142761614</a>

Reviewed by Cameron McCormack.

When calling resize() on a Vector, dynamic offsets from a previous draw
call could remain in the dynamic offsets Vector. Ensure they are zero&apos;ed instead
by using Vector&apos;s ctor instead.

No change needed for RenderBundles since render bundles save all the dynamic
offsets across all draw calls into a single buffer.

* LayoutTests/fast/webgpu/nocrash/fuzz-285828-expected.txt: Added.
* LayoutTests/fast/webgpu/nocrash/fuzz-285828.html: Added.
Add regression test.

* Source/WebGPU/WebGPU/ComputePassEncoder.mm:
(WebGPU::ComputePassEncoder::setPipeline):
* Source/WebGPU/WebGPU/RenderPassEncoder.mm:
(WebGPU::RenderPassEncoder::setPipeline):
Ensure stale dynamic offsets do not persist.

Canonical link: <a href="https://commits.webkit.org/288990@main">https://commits.webkit.org/288990@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/095d32eb429d286be26b1ef9ae0b0086f9337101

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/84873 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/4598 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/39263 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/90017 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/35926 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/4688 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/12575 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/66049 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/23864 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/87918 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/3580 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/77129 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/46319 "Passed tests") | 
| | [⏳ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/iOS-18-Simulator-WPT-WK2-Tests-EWS "Waiting to run tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/31354 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/34999 "Built successfully") | 
| | [⏳ 🧪 api-ios](https://ews-build.webkit.org/#/builders/API-Tests-iOS-Simulator-EWS "Waiting to run tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/32161 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/91390 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/12212 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/8924 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/74538 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/12442 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/72939 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/73663 "Passed tests") | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/18040 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/16486 "Passed tests") | [  ~~🛠 playstation~~](https://ews-build.webkit.org/#/builders/134/builds/4171 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/13243 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/12164 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/17624 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/11999 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/15493 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/13744 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

heycam reviewed (approved):
In the commit message, remove either "should" or "does not". (Actually better is to describe the change, which would be something like "When switching pipelines, zero out dynamic offsets from prior pipeline".)

mwyrzykowski committed:
[WebGPU] Switching pipelines does not zero out dynamic offsets from prior pipeline
https://bugs.webkit.org/show_bug.cgi?id=285828
rdar://142761614

Reviewed by Cameron McCormack.

When calling resize() on a Vector, dynamic offsets from a previous draw
call could remain in the dynamic offsets Vector. Ensure they are zero'ed instead
by using Vector's ctor instead.

No change needed for RenderBundles since render bundles save all the dynamic
offsets across all draw calls into a single buffer.

* LayoutTests/fast/webgpu/nocrash/fuzz-285828-expected.txt: Added.
* LayoutTests/fast/webgpu/nocrash/fuzz-285828.html: Added.
Add regression test.

* Source/WebGPU/WebGPU/ComputePassEncoder.mm:
(WebGPU::ComputePassEncoder::setPipeline):
* Source/WebGPU/WebGPU/RenderPassEncoder.mm:
(WebGPU::RenderPassEncoder::setPipeline):
Ensure stale dynamic offsets do not persist.

Canonical link: https://commits.webkit.org/288990@main
